lsql/judge/forms.py

```diff
@@ -35,6 +35,24 @@ class LoginForm(forms.Form):
     password = forms.CharField(label='Contraseña', max_length=100, widget=forms.PasswordInput)
 
 
+class ResultForm(forms.Form):
+    """Form used results"""
+    group = forms.CharField(label='Grupo', max_length=20)
+    start = forms.DateField(label='Desde', input_formats=['%Y-%m-%d'])
+    end = forms.DateField(label='Hasta', input_formats=['%Y-%m-%d'])
+
+    def clean(self):
+        cleaned_data = super().clean()
+        start = cleaned_data.get("start")
+        end = cleaned_data.get("end")
+        if end is not None and start is not None:
+            if end < start:
+                raise ValueError("Validacion fechas")
```

Comment:
La intención detrás de mis mensajes es usar al máximo las facilidades de Django para validar entradas del usuario, porque eso hace el código más simple, mantenible y homogéneo. Tu commit incorpora muchas de las cosas que hemos comentado pero aún faltarían 3 detalles para explotar adecuadamente las facilidades de Django:

1. Una idea importante de la validación con formularios es usar excepciones ValidationError (ver por ejemplo https://docs.djangoproject.com/en/3.1/ref/forms/validation/#cleaning-and-validating-fields-that-depend-on-each-other para un clean(self)). Estas excepciones se capturan automáticamente en el is_valid y se integran en el atributo form.errors (https://docs.djangoproject.com/en/3.1/ref/forms/api/#django.forms.Form.errors) que tendrá todos los errores encontrados.

2. Si form.is_valid() es falso, directamente generar una página con la plantilla que hiciste mostrando los mensajes de error sacados directamente de form.errors. Eso hace innecesaria la función error_value(err) porque ya no habría excepciones generadas por la validación de entradas y todo se manejaría en exactamente el mismo punto (además de simplificar el flujo de la ejecución). Realmente mi insistencia va dirigida a ese punto: que todo lo que tenga que ver con validación se trate en el formulario y con los mecanismos que proporciona Django.

3. Ahora mismo ResultForm no es aprovechado del todo porque debería poder detectar grupos válidos que serán números naturales (no confundir con grupos que sean números pero no existan, que es otra comprobación diferente de validación). Pero según veo es algo que sigue acabando en error_value, cuando se podría incorporar en ResultForm (ahora mismo es un CharField).

Por favor, dime si hay alguna distinción de casos compleja que impide centralizar toda la validación en ResultForm y generar los mensajes de error a partir de form.errors. Ten en cuenta que se pueden configurar algunos parámetros como opcionales.

Reply:
El punto 3 de mi contestación no tiene sentido, has hecho bien en ignorarlo. Estaba confundiendo grupo, que sí que usamos nombres, con colección, que se identifican con su clave primaria. Perdona la confusión.